Onboarding: Beginner lecture (+ runnable Mesh examples) (#1916)
cardano-foundation/developer-portal: docs/developers/onboarding/lectures/beginner/2-utxos-and-transactions.mdx

Reviewer: I think this lecture overlooks all the onboarding to off-chain SDKs. I think we should either extend to properly introduce off-chain SDKs (what they are, how they work, what they do, how to install and use them), or use a regular self-transaction using the wallet's UI and then do a lecture on off-chain SDKs.

---
title: "UTxOs & transactions"
sidebar_label: "UTxOs & transactions"
description: "How Cardano stores value as sealed bags of coins, and how a transaction moves it."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import extractRegion from "@site/src/utils/extractRegion";
import SendAda from "!!raw-loader!@site/examples/onboarding/lectures/mesh/src/send-ada.ts";

> # UTxOs & transactions

Reviewer: ```suggestion
# UTxOs & Transactions
```
title case

You funded a wallet in the last lecture. Now let's see what that balance really is, and what happens when you spend it.

> On many systems your balance is a single number that goes up and down, like a bank account. Cardano works differently. Your balance is made of separate **sealed bags of coins**. Each bag holds a fixed amount, and once it's sealed **only the blockchain can open it**. Your balance is just the sum of all your bags.

Reviewer: ```suggestion
On many systems, your balance is a single number that goes up and down, like a bank account. Cardano works more similarly to cash. Your balance is made of separate **sealed bags of coins**. Each bag holds a fixed amount of coins determined when creating the bag. Your total balance is just the sum of all the bags in your address.
```

> Each sealed bag is a **UTxO** (an _unspent transaction output_). You might have one bag holding 10 ADA and another holding 2 ADA, together that's a balance of 12 ADA, but they stay two separate bags.

Reviewer: ```suggestion
Each sealed bag is a **UTxO** (an _unspent transaction output_). If your address has one bag holding 10 ADA and another holding 2 ADA, together, that's a balance of 12 ADA, but they stay as two separate bags.
```

> Here's the part that surprises people: **you can't reach into a bag and pull out part of it.** To spend, you hand a whole bag to the blockchain and let _it_ do the opening. Say you have a bag of 10 ADA and want to send 6 to a friend:

Reviewer: ```suggestion
Here's the part that surprises people: **you can't reach into a bag and pull out part of it; they are sealed!** To spend your tokens, you have to do a transaction that releases the tokens by destroying the whole bag (consuming the UTxO), sends the tokens you wanted to spend to the new address, and sends the ones you didn't want to spend back to your address. Both in brand new sealed bags. Say you have a bag of 10 ADA and want to send 6 to a friend. You'll create atransaction that:
```

> 1. You put your **whole 10 ADA bag** into a **transaction**, it goes in as an **input**.

Reviewer: ```suggestion
1. Takes your **whole 10 ADA bag** as an **input**.
2. Destroys the bag, releasing the ADA, to freely rearrange them into **outputs** (new bags).
3. Creates two new sealed bags (outputs), one with 6 ADA on your friend's address and another with 4 ADA in your address.
4. Checks that some rules called "ledger rules" are respected (like the sum of ADA in all inputs is equal to the sum of ADA of all outputs).
```
We shouldn't characterize the transaction as a "thing" that is there, but as something that happens to UTxOs

Author: Agree. I didn't mean the transaction as a "thing," I just used a physical picture (the bag) because it's easier for a newcomer than an abstract state change. The bag/UTxO is the tangible part... the transaction is what happens to it

> 2. The blockchain **opens it**, counts out 6 ADA, and seals them in a **new bag addressed to your friend**.

Reviewer: ```suggestion
```

> 3. It seals the leftover 4 ADA in **another new bag and sends it back to you** as change.

Reviewer: ```suggestion
```

> Those two new bags are the transaction's **outputs**.

Reviewer: ```suggestion
This is a diagram of that transaction. Inputs go in, outputs go out:
```

> ```mermaid

Reviewer: I think we have to make this diagram nicer. A section for addresses, a section for value, a section for name, some notes, etc. Let's explore how to make UTxOs diagrams nicer.

flowchart LR
    I["Your sealed 10 ADA bag<br/>(input)"] --> TX(( Blockchain<br/>opens & re-bags ))
    TX --> O1["new 6 ADA bag<br/>your friend"]
    TX --> O2["new ~4 ADA bag<br/>back to you (change)"]
```

> This is why any amount works: the blockchain isn't handing you an existing "6 ADA note", it **makes fresh bags for the exact amounts**. Your original bag is gone forever, a bag is always used **all at once, never partially** (a small fee is taken from your change). That's the **eUTxO model**, and because every bag going in and coming out is spelled out up front, a Cardano transaction either does exactly what it says or doesn't happen at all.

Reviewer: ```suggestion
This is why any amount works: the blockchain isn't handing you an existing "6 ADA note", it **makes fresh bags for the exact amounts**. Your original bag is gone forever. A bag is always used **all at once, never partially** (a small fee is taken from your change to pay for the processed transaction). That's the **UTxO model**. Each "bag" is called UTxO (_Unspent Transaction Output_) because it's an Output from a previous Transaction that hasn't been spent yet.
```
I think mentioning determinism so briefly could be confusing. We should either explain it properly or not at all. I think not at all right now. Time on Cardano is a better place for this.

That's what a **blockchain** literally is: a chain of blocks, each one linked to the block before it, in order. Once your transaction is inside a block it's **confirmed**, which is why, right after you send, you wait a moment for it to "appear", it's waiting to be picked up into the next block.

```mermaid
flowchart LR
    subgraph Block["Block — a batch of transactions"]
        direction TB
        T1["tx"]
        T2["your tx"]
        T3["tx"]
    end
    Prev["earlier block"] --> Block --> Next["next block"]
```

Think of it like a ledger book: each transaction is one entry, a block is a page that gathers many entries at once, and the pages are bound in order into the book, the chain.

## See it in code

Let's actually create one. This connects your wallet and sends **1 ADA to yourself**, sending to your own address is the simplest way to watch a real transaction appear, with no second wallet to set up. It uses **only the connected wallet**: the wallet gives up the bags to spend, signs, and submits, with no extra service in between.

> Before you read it, one thing worth calling out, because it can feel like magic and it isn't: **you never hand-pick which bag to spend.** Two lines in the code do it for you:

Reviewer: ```suggestion
Before you read it, one thing worth calling out, because it can feel like magic and it isn't: **you don't _have to_ hand-pick which bag to spend if you don't need/want to.** Two lines in the code do it for you:
```

- `await wallet.getUtxos()` asks the wallet for **all** your current bags (UTxOs).

> - `.selectUtxosFrom(...)` hands that whole pile to Mesh, which then does **coin selection**: inside `.complete()` it picks just enough bags to cover your 1 ADA payment plus the network fee, while `.changeAddress(...)` says where the leftover comes back as change.

Reviewer: This code is MeshJS-specific, so it shouldn't live in the generic explanation. We should talk about the requirement that transactions need to be balanced, and that SDKs can do that automatically for us so we don't have to. Then, we can point to the Mesh-specific code: the `complete()` method balances the transaction and uses `selectUtxosFrom()` to choose more utxos if more input value is needed.
